daemons/execd/remoted_schemas.c

```diff
@@ -0,0 +1,248 @@
+            crm_warn("Could not write schema file %s: %s", path, strerror(errno));
+        } else {
+            rc = fprintf(stream, "%s", child->content);
+
+            if (rc < 0) {
+                crm_warn("Could not write schema file %s: %s", path, strerror(errno));
+            }
+
+            fclose(stream);
+        }
+
+        break;
+    }
+
+    free(path);
+}
+
+static void
+get_schema_files(void)
+{
+    int rc = pcmk_rc_ok;
+    cib_t *cib = NULL;
+    xmlNode *reply;
+
+    cib = cib_new();
+    if (cib == NULL) {
+        _exit(ENOTCONN);
+    }
+
+    rc = cib->cmds->signon(cib, crm_system_name, cib_query);
+    if (rc != pcmk_ok) {
+        crm_err("Could not connect to the CIB manager: %s", pcmk_strerror(rc));
+        _exit(pcmk_rc2exitc(rc));
+    }
+
+    rc = cib->cmds->fetch_schemas(cib, &reply, xml_latest_schema(), cib_sync_call);
+    if (rc != pcmk_ok) {
+        crm_err("Could not get schema files: %s", pcmk_strerror(rc));
+        rc = pcmk_legacy2rc(rc);
+
+    } else if (reply->children != NULL) {
+        /* The returned document looks something like this:
+         * <cib_command>
+         *   <cib_calldata>
+         *     <schemas>
+         *       <schema version="pacemaker-1.1">
+         *         <file path="foo-1.1">
+         *         </file>
+         *         <file path="bar-1.1">
+         *         </file>
+         *         ...
+         *       </schema>
+         *       <schema version="pacemaker-1.2">
+         *       </schema>
+         *       ...
+         *     </schemas>
+         *   </cib_calldata>
+         * </cib_command>
+         *
+         * All the <schemas> and <schema> tags are really just there for organizing
+         * the XML a little better.  What we really care about are the <file> nodes,
+         * and specifically the path attributes and the CDATA children (not shown)
+         * of each.  We can use an xpath query to reach down and get all the <file>
+         * nodes at once.
+         *
+         * If we already have the latest schema version, or we asked for one later
+         * than what the cluster supports, we'll get back an empty <schemas> node,
+         * so all this will continue to work.  It just won't do anything.
+         */
+        crm_foreach_xpath_result(reply, "//" PCMK__XA_FILE, write_extra_schema_file, NULL);
+    }
+
+    cib__clean_up_connection(&cib);
+    _exit(pcmk_rc2exitc(rc));
+}
+
+/* Load any additional schema files when the child is finished fetching and
+ * saving them to disk.
+ */
+static void
+get_schema_files_complete(mainloop_child_t *p, pid_t pid, int core, int signo, int exitcode)
+{
+    const char *errmsg = "Could not load additional schema files";
+
+    if ((signo == 0) && (exitcode == 0)) {
+        /* Don't just crm_schema_init here because that will load the base
+         * schemas again too.  Instead just load the things we fetched.
+         */
+        pcmk__load_schemas_from_dir(PCMK__REMOTE_SCHEMA_DIR);
```

Comment:
Shouldn't this check the environment variable first?

Reply:
Which environment variable do you mean?  I didn't introduce a new one for where extra schemas get downloaded - it's just a configure option.

Reply:
Oh, you mean the /etc/sysconfig/pacemaker bit.  Right.